lightning/src/routing/router.rs

```diff
@@ -333,84 +592,262 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+					}
+					if *channel_liquidity_available_msat < spent_on_hop_msat {
+						// This should not happen because we do recompute fees right before,
+						// trying to avoid cases when a hop is not usable due to the fee situation.
+						break 'path_construction;
+					}
+					*channel_liquidity_available_msat -= spent_on_hop_msat;
 				}
+				// Track the total amount all our collected paths allow to send so that we:
+				// - know when to stop looking for more paths
+				// - know which of the hops are useless considering how much more sats we need
+				already_collected_value_msat += value_contribution_msat;
+
+				payment_paths.push(payment_path);
+				found_new_path = true;
+				break 'path_construction;
+			}
 
-				let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
-					Some(hop) => hop.3,
-					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination".to_owned(), action: ErrorAction::IgnoreError}),
-				};
-				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
-				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
-				res.push(new_entry);
+			// Otherwise, since the current target node is not us,
+			// keep "unrolling" the payment graph from payee to payer by
+			// finding a way to reach the current target from the payer side.
+			match network.get_nodes().get(&pubkey) {
+				None => {},
+				Some(node) => {
+					add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node, value_contribution_msat);
+				},
 			}
-			res.last_mut().unwrap().fee_msat = final_value_msat;
-			res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
-			let route = Route { paths: vec![res] };
-			log_trace!(logger, "Got route: {}", log_route!(route));
-			return Ok(route);
 		}
 
-		match network.get_nodes().get(&pubkey) {
-			None => {},
-			Some(node) => {
-				add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
-			},
+		// Step (3).
+		// Stop either when recommended value is reached,
+		// or if during last iteration no new path was found.
+		// In the latter case, making another path finding attempt could not help,
+		// because we deterministically terminate the search due to low liquidity.
+		if already_collected_value_msat >= recommended_value_msat || !found_new_path {
+			break 'paths_collection;
+		}
+	}
+
+	// Step (4).
+	if payment_paths.len() == 0 {
+		return Err(LightningError{err: "Failed to find a path to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	if already_collected_value_msat < final_value_msat {
+		return Err(LightningError{err: "Failed to find a sufficient route to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	// Sort by total fees and take the best paths.
+	payment_paths.sort_by_key(|path| path.get_total_fee_paid_msat());
+	if payment_paths.len() > 50 {
+		payment_paths.truncate(50);
```

Comment:
What if we have 51 payments path, you truncate 50 of them and the one remaining isn't contributing enough ? Maybe keep 50 but truncate by 25 ?

Can you constify the value with a comment explaining his purpose somewhere ? I guess this is covered by our fuzzers, not the functional tests ?

Reply:
>What if we have 51 payments path, you truncate 50 of them and the one remaining isn't contributing enough ?

In this case, it would leave 50 and truncate 1. The parameter 50 says how much should remain.

Spec:
>Shortens the vector, keeping the first len elements and dropping the rest.

Reply:
Lol I should read the docs!